QPID-8484: [Broker-J] Reimplementation of the limit number of connections per user

mklaca opened the pull request.

alex-rufous reviewed (changes requested):
Hi Marek
My apologies for keeping your waiting for the code review. The PR is huge and it took me a while to look through the implementation. I still need to spend some time on UI and test. Thus, these are my intermediate review comments.

For the future, I would prefer to spilt such large PR into parts. Something like
* preparation refactoring (cleaning the code)
* implementation of backend
* implementation of UI
* documentation

That should save both of us a lot of time on implementation and review.

Here is brief summary of my review comments
* IMHO functionality to check limits should live on ConnectionLimitProviders including registering and de-registering connections. On connection registration/deregistration the VH simply needs to iterate through its own and broker ConnectionLimitProviders and call corresponding methods (register/deregister) to verify limits.
* IMHO a special runtime exception can be thrown when any of the limits is breached
* this type of exception can caught in the protocol IO layers and connection can be closed accordingly
* you can potentially use connection delete task to unregister corresponding connection principals on connection close.
* I would prefer to use json/yml over bespoke format. That should reduce implementation code
* The limit rules can be simplified by removing "Blocked". The user simply can create a rule for ALL where connection limit (frequency limit) is 0.
* A frequency value can be set as a number of connections per unit of time in the rules. For example, 1/s,2/m,100/h, 1000/d. I think that would make the rules simpler for the end user.
* the reject reason can be logged as part of connection close operational logs. Thus, new message resources are not really required. We can delete those and all corresponing logging functionality.
* The rule evaluation functionality can be simplified. It looks a bit convoluted at the moment. there some duplicate code across different rule implementation. 

I think that creation a design document (we can use qpid confluence) for this new feature before jumping into implementation would benefit us both and could save us a lot of time. it is still not to late to do it

Kind Regards,
Alex

mklaca commented:
Hi Alex,

> I think that creation of design document (we can use qpid confluence) for this new feature (before jumping into implementation) would benefit us both and could save us a lot of time. it is still not to late to do it.

I don't have rights to create a new or edit any qpid confluence page.

> IMHO functionality to check limits should live on ConnectionLimitProviders including registering and de-registering connections. On connection registration/deregistration the VH simply needs to iterate through its own and broker ConnectionLimitProviders and call corresponding methods (register/deregister) to verify limits. The VH should not call ConnectionLimiter and keep a reference to it. Instead, VH should operate with ConnectionLimitProviders.

If the borker ConnectionLimitProvider kept the ConnectionLimiter then the connection counters inside the limiter would be shared amoung multiple VirtualHosts. The VirtualHosts would effect each other, but it is stated in the documentation - [Java Broker concepts](http://qpid.apache.org/releases/qpid-broker-j-8.0.3/book/Java-Broker-Concepts-Virtualhosts.html) that:
_Virtualhosts are independent; the messaging that goes on within one virtualhost is independent of any messaging that goes on in another virtualhost._
Hence, I proposed the ConnectionLimitProvider as a factory that builds a new independent limiter, which is provided to a VirtualHost.

VirtualHost can not simple iterate through the limit providers or limiters because the consistency of the connection counters is needed. Let's suppose that the VirtualHost has multiple limit providers. If any of the providers/limiters fails to register connection then all have to fail. The connection has to be registered everywhere or nowhere and so the co-ordination is needed. Hence, the virtual host needs a master limiter that co-ordinates the update of the connection counters.

There are concurrency issues with simple iteration. The working thread needs an exclusive access to all connection counters of the user to check and update all counters atomically.

I see three approaches:
1. One connection limiter.
2. Master limiter that co-ordinates all limiters.
3. Chain of responsibility, when the limiter itself call the next limiter.

I have to point out that the functionality of the multiple limiters has not been requested yet.

> IMHO a special runtime exception can be thrown when any of the limits is breached this type of exception can be caught in the protocol IO layers and connection can be closed accordingly.

I have introduced a new ConnectionLimitException but it looks like a checked exception to me.

> I would prefer to use json/yml over bespoke format. That should reduce implementation code.

I want the file based configuration that is backward compatible with existing ACL rule file. Therefore, I had introduced a ACL like bespoke format and so it is not required to generate a new configuration file in different format. The file based configuration can point to the former ACL file.

> The limit rules can be simplified by removing "Blocked". The user simply can create a rule for ALL where connection limit (frequency limit) is 0.

Yes, it can be but 

- Declaring an user as blocked is more friendly and understandable.
- Setting "count_limit=0" could be confusing because the value 0 has meaning "no limit" in many other situations. I don't like the idea of the "magic" constant.
- The blocked user is handled in different way as unblocked user, a connection opened by blocked user can be rejected immediately and any counter is not needed.

> A frequency value can be set as a number of connections per unit of time in the rules. For example, 1/s, 2/m,100/h, 1000/d. I think that would make the rules simpler for the end user.

Yes it is transparent for end user, I have implemented it but the feature complicates evaluation of the rules a lot.

Let suppose that user is a part of the groups gA, gB, gC. Each group has frequency limit:
gA: frequency limit = 10
gB: frequency limit = 5
gC: frequency limit = 15
I can easily pick up the most restrictive limit 5 because all limits have the same time period and I can easily merge all limits into single effective limit.

The frequency limit 1/s is not the same as 60/m, they looks similar but 60/m allows a short peak of new connections, 1/s does not. If multiple rules could have different time period then I can not pick up the most restrictive limit and the user can have multiple frequency limits with different frequency periods.

> The reject reason can be logged as part of connection close operational logs. Thus, new message resources are not really required. We can delete those and all corresponding logging functionality.

We have received complains and our support team queries the logging of the connection limits. Rejecting a new connection because of the breaking the limits is not a regular close operation.

We need:
- Log the limit when a new connection is rejected.
- Log current value of the counter when a new connection is accepted. This should be optional.
- Connection limit messages should be easy to distinguish and filter out.

> A DeleteTask for every identity should be registered on the connection to free the slot.

I can not use DeleteTask because it does not ensure release of the connection slot when any of the delete tasks fails.
Suppose that the connection has 3 delete task, if first task fails then the second and third task are not executed. Hence, the connection counter remains none zero for ever.
The de-registering the connection from VirtualHost is not delete task because of the same reason.

Regards,
Marek

mklaca commented on broker-core/src/main/java/org/apache/qpid/server/logging/messages/UserResourceLimit_logmessages.properties:
```diff
@@ -0,0 +1,23 @@
+# User Resource Limit logging message i18n strings.
+ACCEPTED = URL-1001 : Accepted : {0} {1} by {2} : {3}
+REJECTED = URL-1002 : Rejected : {0} {1} by {2} : {3}
+DEFERRED = URL-1003 : Deferred : {0} {1} by {2}
```
ACCEPTED can not be deleted, we need to log the accepted connection as well with the current connection counter value. Rejecting a new connection because of the breaking the limits is not a regular close operation.

mklaca commented on broker-plugins/user-connection-limits/src/main/java/org/apache/qpid/server/user/connection/limits/config/Limits.java:
```diff
@@ -0,0 +1,70 @@
+
+    static Integer min(final Integer first, final Integer second)
+    {
+        if (first == null)
```
If both rules have only frequency limit then both counter limits will be null. Hence, it is not mistake.

mklaca commented on broker-plugins/user-connection-limits/src/main/java/org/apache/qpid/server/user/connection/limits/config/Limits.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    @Override
+    public boolean isUserBlocked()
```
The blocked user has different treatment than unblocked user. Why is the method redundant ?

mklaca commented on broker-plugins/user-connection-limits/src/main/java/org/apache/qpid/server/user/connection/limits/config/BlockingRule.java:
```diff
@@ -16,29 +16,43 @@
-import org.apache.qpid.server.transport.AMQPConnection;
-
-public class ConnectionPrincipalLimitRule extends ConnectionPrincipalStatisticsRule
+final class BlockingRule extends AbstractRule
```
Blocked user has special treatment and using the magic constant 0 is confusing because 0 means "no limit" in many another situation.

mklaca commented on broker-plugins/amqp-0-10-protocol/src/main/java/org/apache/qpid/server/protocol/v0_10/ServerConnectionDelegate.java:
```diff
@@ -269,8 +270,22 @@ public void connectionOpen(ServerConnection sconn, ConnectionOpen open)
             {
+                final AMQPConnection_0_10 amqpConnection = sconn.getAmqpConnection();
+                final Principal principal = sconn.getAuthorizedPrincipal();
+                if (principal != null)
```
That would be a change of the existing logic, it is not part of this feature. If it was required then another bug task fix should be raised.

mklaca commented on broker-plugins/amqp-0-10-protocol/src/main/java/org/apache/qpid/server/protocol/v0_10/ServerConnection.java:
```diff
@@ -412,13 +410,24 @@ public void closed()
         }
         finally
         {
-            NamedAddressSpace addressSpace = getAddressSpace();
```
If some DeleteTask throws an exception then subsequent DeleteTasks are not executed. Hence, DeleteTask does not ensure that slot is always free-up.

mklaca commented on broker-core/src/main/java/org/apache/qpid/server/virtualhost/UserConnectionLimitChangeListener.java:
```diff
@@ -0,0 +1,78 @@
+import java.util.Objects;
+import java.util.Set;
+
+final class UserConnectionLimitChangeListener extends AbstractConfigurationChangeListener
```
In case of multiple connection limiters every virtual host needs a master coordinator that keeps the consistency of the counters inside the connection limiters. Hence, the functionality can not be moved. The master coordinator has to be notified about the change.

mklaca commented on broker-core/src/main/java/org/apache/qpid/server/virtualhost/AbstractVirtualHost.java:
```diff
@@ -2691,6 +2727,18 @@ public String getArguments()
     }
 
+    @Override
+    public ResourceAvailability requestConnectionSlot(AMQPConnection<?> connection, String userId)
```
This leads to the sharing of the connection counters among VistualHosts. But, it is stated in documentation that VistualHosts are independent, hence every VistualHosts needs it's own ConnectionLimiter with independent connection counters.

mklaca commented on broker-core/src/main/java/org/apache/qpid/server/virtualhost/AbstractVirtualHost.java:
```diff
@@ -320,6 +323,8 @@ public AbstractVirtualHost(final Map<String, Object> attributes, VirtualHostNode
 
         _fileSystemSpaceChecker = new FileSystemSpaceChecker();
+        _userConnectionLimiter = UserConnectionLimiter.noLimits();
+        _userConnectionLimitChangeListener = new UserConnectionLimitChangeListener(this);
```
The "this" pointer should not be used before finished construction. But this technique is already used multiple times in VirtualHost class and the fix requires significant refactoring. If you wish to fix it raise a separate bug fix please.

mklaca commented on broker-core/src/main/java/org/apache/qpid/server/virtualhost/AbstractVirtualHost.java:
```diff
@@ -229,6 +233,7 @@ public Result getResult(final Subject subject)
         }
     }, Result.DEFER);
 
+    private volatile UserConnectionLimiter _userConnectionLimiter;
```
Every VH needs a master connection limiter that will co-ordinate the work of all connection limiters and keep the connection counters in consistent state.

mklaca commented on broker-core/src/main/java/org/apache/qpid/server/transport/AbstractAMQPConnection.java:
```diff
@@ -1161,28 +1161,35 @@ public void incrementTransactionBeginCounter()
 
-    @Override
-    public int getAuthenticatedPrincipalConnectionFrequency()
+    private final class ConnectionSlot
```
As per comments above a DeleteTask can not do the clean-up.

mklaca commented on broker-core/src/main/java/org/apache/qpid/server/transport/AbstractAMQPConnection.java:
```diff
@@ -1161,28 +1161,35 @@ public void incrementTransactionBeginCounter()
 
     @Override
-    public void registered(final ConnectionPrincipalStatistics connectionPrincipalStatistics)
+    public void freeConnectionSlots()
```
If some DeleteTask throws an exception then subsequent DeleteTasks are not executed. Hence, a new DeleteTask will not ensure that slot is always released and can not be used to free connection slot.

mklaca commented on broker-core/src/main/java/org/apache/qpid/server/transport/AMQPConnection.java:
```diff
@@ -148,10 +147,7 @@ void registerTransactionTickers(ServerTransaction serverTransaction,
-    int getAuthenticatedPrincipalConnectionCount();
-
-    int getAuthenticatedPrincipalConnectionFrequency();
+    void addConnectionSlot(NamedAddressSpace addressSpace, String userId);
```
If some DeleteTask throws an exception then subsequent DeleteTasks are not executed. Hence, a new DeleteTask will not ensure that slot is always released and can not be used to free connection slot.

mklaca commented on broker-core/src/main/java/org/apache/qpid/server/security/UserConnectionLimitProvider.java:
```diff
@@ -0,0 +1,32 @@
+@FunctionalInterface
+public interface UserConnectionLimitProvider
+{
+    UserConnectionLimiter getLimiter(EventLoggerProvider eventLoggerProvider);
```
This leads to the sharing of the connection counters among VistualHosts, because broker connection limit providers are common for all  VistualHosts. But, it is stated in documentation that VistualHosts are independent, hence every VistualHosts needs it's own ConnectionLimiter with independent connection counters. ConnectionLimitProvider works as factory that build a new independent limiter for each VirtualHost.

mklaca commented on broker-core/src/main/java/org/apache/qpid/server/model/port/AmqpPortImpl.java:
```diff
@@ -542,69 +539,64 @@ public static String getInstalledProtocolsAsString()
-                                                PortMessages.CONNECTION_REJECTED_TOO_MANY(addressString,
-                                                                                       _maxOpenConnections));
-            return false;
+            final long currentCount = _connectionCount.getAndUpdate(count -> count < maxOpenConnections ? count + 1L : count) + 1L;
```
Every thread has it's own independent stack, hence local variables are not shared among threads.

mklaca commented on broker-core/src/main/java/org/apache/qpid/server/model/NamedAddressSpace.java:
```diff
@@ -48,8 +50,17 @@
+
     void deregisterConnection(AMQPConnection<?> connection);
 
+    default ResourceAvailability requestConnectionSlot(AMQPConnection<?> connection, String userId)
```
This leads to the sharing of the connection counters among VistualHosts, because broker connection limit providers are common for all VistualHosts. But, it is stated in documentation that VistualHosts are independent, hence every VistualHosts needs it's own ConnectionLimiter with independent connection counters. ConnectionLimitProvider works as factory that build a new independent limiter for each VirtualHost.

mklaca commented:
Hi @alex-rufous,

do you have any queries, doubts, comments or suggestions?

Regards,
Marek

alex-rufous commented on broker-core/src/main/java/org/apache/qpid/server/virtualhost/VirtualHostConnectionLimiter.java:
```diff
@@ -0,0 +1,294 @@
+        }
+    }
+
+    private static final class VirtualHostChangeListener extends AbstractChangeListener
```
The VirtualHostChangeListener code is a duplicate of BrokerChangeListener.
It seems it can be parametrised using category class, and child category class. Those can be passed into constructor

alex-rufous reviewed (changes requested):
Hi Marec,
As per our discussion I reviewed the changes submitted by you for re-implementation of connection limiting functionality.
I did a very thorough review of  the changes implemented in broker-core and I gave a very brief  look into implementation  details and UI.

I think I am OK with the implementation code submitted in qpid-broker-plugins-user-connection-limits, though, I am going to review it a bit more details later.

So far, I would like to request changes in the broker-core.

As per my review comments I would like you to change code in connection implementations and some base classes and implementation.
For a simplicity I just copied  my essential review comments below

* ConnectionSlot
As far as I understood, the responsibility of this interface is provide a mechanism to de-register a connection on the connection limiter when the connection is closed.
There is already an existing mechanism to perform any pending close/delete activities which is  a "delete task".
Please use the existing mechanism. There is no need to introduce a new one.
If an existing mechanism has flows, those flows has to be fixed. For example, if a throwing an exception as part of delete task breaks the unregistration logic, please add exception handling into execution of "delete tasks" to make sure that all of them get executed regadless failing ones. If required, the delete task exceptions could be re-thrown after execution all of them.
Thus, I strongly believe that ConnectionSlot interface is redundant interface which duplicates existing "delete task" functionality. Please delete ConnectionSlot.

* AMQPConnection

Please remove new methods in AMQPConnection interface:

** void addConnectionSlot(NamedAddressSpace addressSpace, String userId) throws ConnectionLimitException;
** void freeConnectionSlots();
    
As per  review comments for ConnectionSlot the "delete task" should be used to perform deregistration activities. The introduced methods addConnectionSlot, freeConnectionSlots duplicate the "delete task" functionality. 

* AbstractAMQPConnection

Please remove _registeredConnectionSlots
As per  review comments for ConnectionSlot, the "delete task" should be used to perform deregistration activities. 
It seems the  field _registeredConnectionSlots  duplicates filed _connectionCloseTaskList

public void freeConnectionSlots()
Please remove freeConnectionSlots. It duplicates a functionality provided by Deletable#addDeleteTask(Action<?>)
You can add an exception handling into functionality to execute delete tasks in order to make sure that all them are invoked. That would ensure that the connection is always deregistered


Please remove addConnectionSlot.
As per comments above, method #addDeleteTask needs to be used in order to create a callback for deregistering of connection on a limiter.


* NamedAddressSpace

** ConnectionSlot requestConnectionSlot(AMQPConnection<?> connection, String userId)

At the moment only authenticated principal is passed into the method to check the limits. What about group principals? if limit is set for entire group, how groups are supposed to be checked?

Does it make sense to check also for group principals in addition to authenticated principal?
Please note that there is no reason to pass userId. The AMQPConnection comes with methods to get authorized principal( getAuthorizedPrincipal ) and subject (getSubject).

I would like to suggest a removal of userId parameter.

The method needs to return void. (Potentially boolean can be returned)

I think that method name should be renamed into checkUserConnectionLimit or a bit more appropriate name. I think that that the name requestConnectionSlot does not represent what method does.


* AmqpPortImpl
** public boolean acceptNewConnectionAndIncrementCount(final SocketAddress remoteSocketAddress)

The implemented change does not belong to the scope of this PR. IMHO, it is a separate change. if there is an issue with a counter it has to be changed in its own JIRA. 

The current change consists of  atomic and non atomic operation. That is an always a scenario for a disaster in a multi-threaded environment.

To be fair I would rather have the code changed to something like the one below 

    boolean limitReached = false
    for (;;) {
            long current = _connectionCount.get();
            limitReached = current >= maxOpenConnections; 
            long newValue = limitReached ?  count : count + 1L
            if (_connectionCount.compareAndSet(current, newValue))
                break;
    }

Though, I am not sure whether even the code above is correct. The original implementation expected that the limit is increased on connection open and decreased on connection close. Thus, we need to check that the close functionality properly decrements the limit.

It could be even safer to go with something like below

    long  currentCount = _connectionCount.incrementAndGet();
    if (currentCount > maxOpenConnections)
    {
        _container.getEventLogger().message(....);
        _connectionCount.decreementAndGet();
       return false;
   }

I would really prefer to create a separate PR for the change.


* CachedConnectionLimiterImpl


Could you please expand why CachedConnectionLimiterImpl is required?
What is the use case for this implementation?

Does it mean that the same CachedLimiter can be called multiple times? If so, why the implementation allow that?


* ConnectionLimiter

** ConnectionSlot register(AMQPConnection<?> connection);
As per comments on ConnectionSlot, the interface ConnectionSlot seems redundant to me.
Thus, after deletion of ConnectionSlot, the method should return void.

** ConnectionLimiter append(ConnectionLimiter limiter);
What are the responsibility of this method?
How many underlying implementations can be appended?
What happens when append is called multiple times? How the appended multiple ConnectionLimiter behaves? 
if the intention here to call the method once, the name should be changed to setUnderlyingConnectionLimiter
What object is responsible for calling append method?
Why this method is actually required on the ConnectionLimiter interface?


** cacheConnectionRegistration
The implementation is really confusing.
Both NoLimits and BlockEveryone are CachedLimiter.
If those would be passed here, the real  CachedConnectionLimiterImpl will not be created.
That's for me looks like a defect in the implementation.
Why this method is required?

** ConnectionLimiter.NoLimits
Why NoLimits implements CachedLimiter? It does not do any caching...

** NoLimits.append
I would like to understand what are the reasons behind such implementation?
The implemented behaviour looks curious. The output depends on the input... Is it correct?


** ConnectionLimiter.BlockEveryone
Why BlockEveryone implements CachedLimiter? It does not do any caching...

** BlockEveryone.underlyingLimiter
The implementation indicates that BlockEveryone should not really implement CachedLimiter

* ConnectionLimiterContext

Why ConnectionLimiterContext is a configured object?
What is the purpose of this interface?
IMHO, the context variable should be defined on the ConnectionLimiterProvider
Why we cannot merger this interface into BrokerConnectionLimiter?


* AbstractVirtualHost
** public ConnectionSlot requestConnectionSlot(AMQPConnection<?> connection)

I am still not convinced that implementation of VirtualHostConnectionLimiter brings any benefits here.
IMHO, from my point of view the iteration over VirtualHost and Broker ConnectionLimitProviders will do absolutely the same job as VirtualHostConnectionLimiter with a complex ecosystem of ConfigurationObjectListeners.

Here is what I believe the implementation of requestConnectionSlot should look like without VirtualHostConnectionLimiter (the code is for illustration purpose; I did not pay much attention on syntax)


    @Override
    public ConnectionSlot requestConnectionSlot(AMQPConnection<?> connection)
    {
        final Collection<VirtualHostConnectionLimitProvider> hostLimiters = _virtualHost.getChildren(VirtualHostConnectionLimitProvider.class);
        final Collection<BrokerConnectionLimitProvider> brokerLimiters = _broker.getChildren(BrokerConnectionLimitProvider.class);
        final List<? extends ConnectionLimitProvider> allLimiters = new ArrayList<>()
        allLimiters.addAll(hostLimiters);
        allLimiters.addAll(brokerLimiters);

	boolean success = false;
	try
	{
		for (ConnectionLimitProvider limiter : allLimiters)
		{
		    limiter.registerAndCheckLimit(connection);
		}
		success = true;
	}
	finally
	{
		if (success)
		{
			connection.addDeletedTask(()-> {
				 final Collection<VirtualHostConnectionLimitProvider> hostLimiters = _virtualHost.getChildren(VirtualHostConnectionLimitProvider.class);
        			 final Collection<BrokerConnectionLimitProvider> brokerLimiters = _broker.getChildren(BrokerConnectionLimitProvider.class);
        			 final List<? extends ConnectionLimitProvider> allLimiters = new ArrayList<>()
				 allLimiters.addAll(hostLimiters);
				 allLimiters.addAll(brokerLimiters);
        			 for (ConnectionLimitProvider limiter : limiters)
				 {
				    limiter.deregister(connection);
				 }
			
			})
		}
		else
	        {
		   for (ConnectionLimitProvider limiter : allLimiters)
		   {
		       limiter.deregister(connection);
		   }
		
		}
	}
    }

The advantage of the above is that you do not need to keep complex relationship between the ConnectionLimitProviders. Thus, the code to maintain is reduced significantly...
The ConnectionLimitProvider interface would be very simple:

    interface ConnectionLimitProvider {

       void registerAndCheckLimit(AMQPConnection<?> connection);
       void deregister(AMQPConnection<?> connection);
    }

If required we can introduce priority to call ConnectionLimitProviders in a priority order.

The disadvantage of such approach that each implementation ConnectionLimitProvider would have to maintain and manage  connections  separately.

mklaca commented on broker-core/src/main/java/org/apache/qpid/server/model/NamedAddressSpace.java:
```diff
@@ -48,8 +49,13 @@
+
     void deregisterConnection(AMQPConnection<?> connection);
 
+    default ConnectionSlot requestConnectionSlot(AMQPConnection<?> connection)
```
ConnectionSlot is a crucial part of the solution, see my last comment.

mklaca commented on broker-core/src/main/java/org/apache/qpid/server/security/limit/ConnectionSlot.java:
```diff
@@ -0,0 +1,75 @@
+import java.util.Optional;
+
+@FunctionalInterface
+public interface ConnectionSlot extends Runnable
```
ConnectionSlot is a crucial part of the solution, it can be used as delete task.

mklaca commented on broker-core/src/main/java/org/apache/qpid/server/security/limit/ConnectionLimiter.java:
```diff
@@ -0,0 +1,114 @@
+        }
+    }
+
+    final class NoLimits implements CachedLimiter
```
Because NoLimits  is a CachedLimiter.

mklaca commented on broker-core/src/main/java/org/apache/qpid/server/security/limit/ConnectionLimiter.java:
```diff
@@ -0,0 +1,114 @@
+        }
+    }
+
+    final class BlockEveryone implements CachedLimiter
```
Because BlockEveryone is a CachedLimiter.

mklaca commented on broker-core/src/main/java/org/apache/qpid/server/virtualhost/AbstractVirtualHost.java:
```diff
@@ -2691,6 +2704,12 @@ public String getArguments()
+    @Override
+    public ConnectionSlot requestConnectionSlot(AMQPConnection<?> connection)
+    {
+        return _connectionLimiter.register(connection);
```
See my last comment.

mklaca commented:
Hi @alex-rufous ,

your suggested implementation of the requestConnectionSlot suffers from multiple issues, it is difficult to fix them all.
The simple iteration over VirtualHost and Broker ConnectionLimitProviders does not do the same job as VirtualHostConnectionLimiter.

```
@Override
public ConnectionSlot requestConnectionSlot(AMQPConnection<?> connection)
{
    final Collection<VirtualHostConnectionLimitProvider> hostLimiters = _virtualHost.getChildren(VirtualHostConnectionLimitProvider.class);
    final Collection<BrokerConnectionLimitProvider> brokerLimiters = _broker.getChildren(BrokerConnectionLimitProvider.class);
    final List<? extends ConnectionLimitProvider> allLimiters = new ArrayList<>()
    allLimiters.addAll(hostLimiters);
    allLimiters.addAll(brokerLimiters);
    
    boolean success = false;
    try
    {
        for (ConnectionLimitProvider limiter : allLimiters)
        {
            limiter.registerAndCheckLimit(connection);
        }
        success = true;
    }
    finally
    {
        if (success)
        {
            connection.addDeletedTask(()-> {
                final Collection<VirtualHostConnectionLimitProvider> hostLimiters = _virtualHost.getChildren(VirtualHostConnectionLimitProvider.class);
                final Collection<BrokerConnectionLimitProvider> brokerLimiters = _broker.getChildren(BrokerConnectionLimitProvider.class);
                final List<? extends ConnectionLimitProvider> allLimiters = new ArrayList<>()
                allLimiters.addAll(hostLimiters);
                allLimiters.addAll(brokerLimiters);
                for (ConnectionLimitProvider limiter : limiters)
                {
                    limiter.deregister(connection);
                }
            })
        }
        else
        {
            for (ConnectionLimitProvider limiter : allLimiters)
            {
                limiter.deregister(connection);
            }
        }
    }
}
```

I have to notice that 'decrement' and 'revert' are not  same operations in case of frequency limit. You need a special 'revert' method to handle exceptions properly.

### Issues:

- **The issues connected with a configuration change:**
What does happen in following scenario?
1. 'requestConnectionSlot' collects all limiters.
2. The connection is registered in all collected limiters, it executes register method on every limiter.
3. The user changes the configuration of the broker.
4. The delete task collects all limiters.
5. The connection is de-registered in all collected limiters.

**Examples:**

- There are two limiters (LimiterA, LimiterB) and then the user changes the configuration, he adds a new LimiterC. The connection was register only in LimiterA and LimiterB and so it has to be deregistered only in LimiterA and LimiterB. Your proposed solution will also de-register the connection in the LimiterC and that is a bug.
- There is just single limiter but the user decides to reset the counters, the old counter inside the limiter is dropped and a new one is created. 'requestConnectionSlot' increments the old counter but the new incorrect limiter is decremented in the delete task.

The connection could remember the limiters where it was registered but that is memory consuming. If the user deleted a limiter then the limiter would remain stored inside the connection.

- **The consistency issue:**
Suppose three limiters LimiterA, LimiterB, LimiterC and the following scenario:
1. The method 'requestConnectionSlot' collects all limiters.
2. The connection is registered in LimiterA.
3. The registration of the connection in LimiterB fails and an exception is thrown.
4. The connection is de-registered in LimiterA, LimiterB, LimiterC. That is a bug, connection was registered only in LimiterA and so the change has to be reverted only in LimiterA.

You need a special 'revert' method that properly reverts the changes and checks that the connection was registered in the limiter indeed.

- **The concurrency issues:**
Every connection should be register in all limiters or it should not be registered in any limiter. But the suggested implementation allows a lot of intermediate states when the connection is 'half' registered or 'half' de-registered. Let have a limiter with frequency limit: one connection per minute.
Suppose following scenario:
1. A new connection is open at 7:01:00 for user M. It is the first connection and so it should pass.
2. A new connection is open at 7:01:59 for user M. It should be rejected.
3. A new connection is open at 7:02:01 for user M. It should pass. But there could be multiple limiters and the result depends on how fast can the second connection revert its changes.

A shared common lock is required that ensures the atomicity of the manipulation with counters. The lock has to ensure that the third connection is waiting till the second finishes its work.

The simple iteration is not a feasible solution. If we fixed all mentioned issues then the solution would become complicated and cluttered and the limiter would need a new methods "deregister" and "revert". The solution requires also significant changes in the exiting code of AMQPConnection_1_0 and other classes.

**My solution:**
The 'append' method creates a new chained limiter without any impact on the input limiters. It does not have any side effect and so the 'append' operation can be repeated.
`ConnectionLimiter chainedLimiter = limiterA.append(limiterB). // The operation does not change neither limiterA nor limiterB. The chainedLimiter is a new limiter that contains limiterA and limiterB.`
The chained limiter ensures proper locking and avoids the concurrency issues. Every user has his own independent lock. Hence, multiple connections from different users can be processed in parallel. The append operation does not have any impact on input limiters. Therefore, it is possible to build a new chained limiter while the old limiter stays untouched and in use. It means easy update when the configuration is changed.

The chained limiter takes care of the consistency issue. The registration is splitted into three steps:
1. The user's lock is requested at first.
2. All counters are check.
3. If only all checks pass then all counters will be incremented.

There is no need for a revert method. The register method returns a connection slot, the connection slot points to a counter object. It remembers which counter should be decremented when the connection slot is released.
The connection slots can be chained and so the chained limiter returns a chained connection slot. The connection slot remembers correct counters regardless of the subsequent configuration change. Therefore, it effectively solves problems caused by any configuration change and neither 'deregister' nor 'revert' method of the limiter are needed.

The cached limiter returns the same connection slot when the same connection is registered multiple times, AMQPConnection_1_0 can try to register itself multiple times. Cached limiter associates the connection with a slot and the following condition is held:
`cachedLimiter.register(connection) == cachedLimiter.register(connection)`
The connection slot produced with cached limiter is released only once regardless how many times the method 'free' is executed.

All user limiting logic is focused in implementations of ConnectionLimiter interface and limiters take care of synchronization, consistency, configuration changes.

**BlockEveryone limiter:**
Let wrap 'BlockEveryone' limiter in cache decorator that utilizes a map as the internal cache. But the cache remains empty forever because the register method of 'BlockEveryone' limiter always throws an exception. There is not any difference in acting of wrapped and unwrapped 'BlockEveryone' limiter.
'BlockEveryone' limiter acts as cached limiter, hence it is a cached limiter.

**NoLimits limiter:**
Let wrap 'NoLimits' limiter in cache decorator that utilizes a map as the internal cache. But the cache will contain the same response FreeSlot.INSTANCE for every input and so it is useless. The wrapped and unwrapped 'NoLimits' limiter returns the same response for every input.
Hence, 'NoLimits' limiter acts as cached limiter and so it is a cached limiter.

alex-rufous commented:
Hi Marek,
Thanks for making the changes and  detailed answers on my questions and comments. 
I think I am happy with the implementation. Could you please merge all your commits into a single commit?
I will merge the changes into main.

Kind Regards,
Alex

mklaca commented:
Hello @alex-rufous, I squashed all commits to one.
